--- a/ncurses/base/lib_mouse.c
+++ b/ncurses/base/lib_mouse.c
@@ -389,11 +389,11 @@ init_xterm_mouse(SCREEN *sp)
     sp->_mouse_format = MF_X10;
     sp->_mouse_xtermcap = tigetstr(UserCap(XM));
     if (VALID_STRING(sp->_mouse_xtermcap)) {
-	char *code = strstr(sp->_mouse_xtermcap, "[?");
+	const char *code = strstr(sp->_mouse_xtermcap, "[?");
 	if (code != NULL) {
 	    code += 2;
 	    while ((*code >= '0') && (*code <= '9')) {
-		char *next = code;
+		const char *next = code;
 		while ((*next >= '0') && (*next <= '9')) {
 		    ++next;
 		}
